### ISSUE TITLE
RFC: The Undefined a.k.a. dealing with unset fields

### ISSUE THREAD

**Issue by bintoro** [schematics/schematics#263]
Everybody probably agrees that Schematics needs the ability to distinguish between fields that have been set to `None` and fields that have not been assigned any value. The issue pops up all the time in various guises, recently as #249.

I have written a patch that will resolve this once and for all. It's available for testing at https://github.com/bintoro/schematics/tree/undef-value-v1.

Diff vs master: https://github.com/bintoro/schematics/compare/undef-value-v1

Documentation: https://gist.github.com/bintoro/c767f314fe827d8082a2

I posted on the dev list a while back, but since the list is so dead, let's try again here. Also, since then, I've been able to simplify the code and documentation quite a bit by removing a certain needlessly complicated feature.

All kinds of feedback would be appreciated, for example:

*   How to deal with `required` fields? Should the `required` parameter be changed to be able to assume three different values representing "not required", "required", and "something better than `None` required"?

*   What does the upgrade path look like? This is a breaking change of sorts, so it's unfortunate that v1.0 is already out of the oven. If you ran the code, how did *your* application deal with the change in the meaning of `obj.somefield is None`? Should the feature ship as disabled by default until some future release?

*   Is there actually any use case for serializing `Undefined` as anything other than `None`? If yes, are the presently available hooks sufficient?

**Comment by bintoro** [comment]
Okay guys, one more alternative, and now this one is fully backward-compatible.

The idea is to make a field's "definedness" a property that must be specifically queried. Upon regular attribute access `None` would be returned for both undefined and null fields. Along these lines:

```python
>>> instance = SomeModel()
>>> print(instance.somefield)
None
>>> isset(instance.somefield)
False
>>> instance.somefield = 1
>>> isset(instance.somefield)
True
>>> instance.somefield = None
>>> isset(instance.somefield)
True
>>> del instance.somefield
>>> isset(instance.somefield)
False
```

Under the earlier proposals, the statement ``print(instance.somefield)`` would have either printed ``Undefined`` or raised an error.

The ``atoms()`` function in ``transforms.py`` would be supplanted by an API that lets application code choose whether unset fields should be represented as ``None`` or ``Undefined`` objects during iteration.

**Comment by lkraider** [comment]
Has this landed into mainline yet?

**Comment by bintoro** [comment]
Okay, so this thing was left in limbo because no one ever provided any feedback, and then I had to step away for a while. Here's a summary of the issue so far. I'll try to keep this as brief as possible while including enough details.

Around the time I started this thread, I identified three strategies to distinguish between `None` and "not set". Eventually, I realized that the initial approach (visible `Undefined` objects) was a bad idea because the expression `x is None` would counterintuitively evaluate to `False` when `x == Undefined`. A lot of existing code would silently break.

That left two alternatives:

### Pythonic strategy ###

From an application's point of view, **unset fields would behave much like non-existent attributes on any regular Python object**. (Internally, though, `_data` might be populated with `Undefined` objects for bookkeeping.)

```python
class Foo(Model):
    somefield = StringType()
```

```python
>>> instance = Foo()
>>> instance.somefield
AttributeError: 'Foo' object has no attribute 'somefield'
```

To assist with backward-compatibility, there would be an easy way to auto-initialize empty fields to `None` so as to match current behavior. Strategies (not mutually exclusive) include global or model-level options, a `Model.__init___` argument, and a dedicated method to populate an instance's empty fields at will.

To check if a field is set, you'd use `hasattr(instance, 'somefield')`, as with any object, or the dict-like interface `'somefield' in instance`, which is already supported.

`Model.get(fieldname)` would return a field's value or `None` if the field is empty. So, in cases where you don't care about the unset/`None` distinction, `instance.get('somefield')` would be equivalent to `instance.somefield` today.

### Tack-on strategy ###

Superficially, things would work like now. `None` would still be returned when an unset field is accessed, but **an application could explicitly check if a `None` actually means `None` or if it means "not set".**

```python
>>> instance = Foo()
>>> print(instance.somefield)
None
>>> isset(instance, 'somefield')
False
```

```python
>>> instance.somefield = 1
>>> isset(instance, 'somefield')
True
>>> instance.somefield = None
>>> isset(instance, 'somefield')
True
```

```python
>>> del instance.somefield
>>> isset(instance, 'somefield')
False
```

(The above examples use a global `isset` function, but it could be a `Model` method as well.)

### Evaluation ###

Personally I would be fine with either alternative.

The tack-on strategy is appealing in that it seems to be 100% backward-compatible, with the obvious exception of code that manipulates internals like `_data` directly. Regular applications can start using the feature right away with no changes to existing code (as far as I can work out).

The pythonic strategy can also be implemented selectively in an existing codebase using the compatibility helpers I mentioned. However, once you have a model that follows the new paradigm, you can't put it through code that blindly accesses the model's fields without checking first (or using `.get()`), as that may raise an `AttributeError`.

Thoughts?

Once we have even a tentative decision, I can deliver a draft PR pretty quickly. If no one comments, then I'll decide. :grin:

**Comment by lkraider** [comment]
Regarding the Pythonic, the exception should probably have a custom message to help when debugging code. Maybe use a custom Exception extending from AttributeError?
```python
>>> instance = Foo()
>>> instance.somefield
ModelFieldError: 'Foo' object has uninitialized field 'somefield'
```

In either of the proposals, how is the output of undefined fields handled? Can implementations redefine what the output will be when serializing (maybe subclassing Undefined and setting it on the field)?

**Comment by bintoro** [comment]
> Maybe use a custom `Exception` extending from `AttributeError`?

Mm-hmm, good idea.

> In either of the proposals, how is the output of undefined fields handled?

The `serialize_when_undefined` boolean (analogous to `serialize_when_none`) determines whether unset fields are exported at all. If they are, the default representation is `None`.

To output something beyond these alternatives, custom logic must be specified.

Subclassing `Undefined` was the route I took first, but it felt like a heavy-handed approach to functionality that would likely be used in fairly simple ways.

I changed it to a function-based system that's [described in the documentation](https://gist.github.com/bintoro/c767f314fe827d8082a2#customizing) of the now-rejected first iteration.

But I don't know, maybe some combination of the two would be best. The newer approach doesn't lend itself well to a situation where you have a set of rules that appear in several places but are not quite universal.

It would help if people could post their real-world scenarios where custom serialization rules are needed. I have a hard time coming up with examples, because if my data has empty fields at export time, I mostly want them to stay that way.

**Comment by lkraider** [comment]
I don't have real examples either, but I like the idea of each type having the ability to define its UndefinedType, just so the model doesn't require a large switch/case method testing all possible types.

Anyway, my vote is for the Pythonic option.

**Comment by lkraider** [comment]
I prefer to err on the side of it being more flexible, specially about it being extendable in runtime and the single function in the Model is problematic for that.

But yes, I agree in releasing this feature now and then extending customization later.

**Comment by bintoro** [comment]
The `required` and `serialize_when_none` settings need to be expanded or complemented so that users can express preferences for both `None` and `Undefined`. Opinions welcome.

### `required`

Available configurations:

  * (A) Value must be present and must be other than `None`.
  * (B) Value must be present but may be `None`.
  * (C) No value required.

Internally, this will likely become a three-mode option, e.g. `False | 'any' | 'nonempty'`. (Could be either strings or constants.)

In the API, one way to do this is to keep using `required` for A and introduce a separate boolean to express B, for example `required_data=True`. The two switches wouldn't have to appear together, so this wouldn't bloat field configs.

An alternative is to expose the three modes directly under the existing `required` option.

Since the `'nonempty'` case is what people usually want, and `required=True` could remain as a synonym for `required='nonempty'`, this repurposing would be backward-compatible. The expression `bool(somefield.required)` can be made to retain its current semantics by turning `BaseType.required` into a property.

### `serialize_when_none`

Available configurations:

  * (A) Export empty (`None/[]/{}`) and missing fields (the default).
  * (B) Export empty fields but not missing fields.
  * (C) Don't export either.

Unlike with `required`, hosting all three options under the existing switch is not workable.

So, the choice is between introducing a parallel setting (`serialize_when_undefined`) or deprecate `serialize_when_none` and replace it with a three-mode setting with a more suitable name.

I'm leaning towards a completely new setting here because
  * the two-setting system seems contrived and might be confusing
  * sometimes (although rarely) you'd have to specify both settings on the same field
  * the naming of `serialize_when_none` is less than ideal, since it actually means "export when empty".

There are two opposite ways to implement the setting:

  * what to export, e.g. `export = 'always' | 'any' | 'nonempty'`
  * what to filter out, e.g. `export_filter = False | 'undefined' | 'empty'`

Whatever the design, the semantics should also make sense when used at export time: `obj.to_primitive(filter='empty')`.

**Comment by bintoro** [comment]
The undefined feature is now for the most part ready, but I'm having trouble deciding on the appropriate default behavior.

There will be two new import-stage options:

  * `apply_defaults` (described in detail in #360)
  * `init_values`, which controls whether to initialize values to `None` or leave them out (`Undefined`).

By default, neither gets an explicit value, in which case the effective values depend on `partial`:

  * `partial=False` implies `True` for both `apply_defaults` and `init_values`.
  * `partial=True` implies `False` for both `apply_defaults` and `init_values`.

To me, this seems pretty natural: if you want a full instance, every field will have a value, be it `None` or a field-specific default; if you want to work with partial data, nothing is set unless present in the input.

Here's the predicament: the current default for `Model.__init__()` is `partial=True`. Considering the above, I feel `partial=False` might be a more natural default.

Does anyone know if there's a particular reason `partial=True` is the default, or was it more or less a random choice?

If we retain `partial=True` as the default, doing `m = SomeModel()` would no longer set any values on the newly created instance, which can seem surprising. Accessing any field right away would raise a `MissingValueError`.

On the other hand, if the default is changed to `partial=False`, doing `m = SomeModel()` would populate all fields with something.

I'm tentatively in favor of changing the default. If we go that route, I think it would make sense for `import_loop` not to complain about required fields when no input data has been provided:

```python
class SomeModel(Model):
    f = StringType(required=True)
```
```python
>>> m = SomeModel()
>>>
```
```python
>>> m = SomeModel({})
  ...
schematics.exceptions.ModelConversionError: {'f': [u'This field is required.']}
```

Any comments?

**Comment by bintoro** [comment]
Okay, I'm not so sure anymore. The tests are full of cases like this:

```python
class User(Model):
    name = StringType(required=True)
    age = IntType(required=True)

u = User({'name': u'Henry VIII'})
u.age = 99
u.validate()
```

If `partial=False` is going to be the new default, the model instantiation would have to say:

```python
u = User({'name': u'Henry VIII'}, partial=True)
```

Admittedly it's pretty convenient to be able to create incomplete instances just like that, without always adding `partial=True`.

If we want to keep it that way, a couple of options come to mind:

  * (A) Keep `partial` completely independent of `apply_defaults` and `init_values`.

    I don't like this because of the amount of complexity it adds. Semantically, `partial` would be ideal
    as a shortcut for setting the new options.

  * (B) Stop enforcing `required` in `import_loop` altogether and move it to the validators.

    I like this idea. Whether a field is optional or required is more like a validation concern than a conversion concern anyway. During validation, the `partial` parameter would continue to govern whether `required` is enforced, but during conversion this would no longer happen.

**Comment by edc** [comment]
A possible solution is to add a new classmethod for instantiation:

```python
u = User()
u.age = 99
print u.age  # => 99
print u.name  # => None

u = User.create()  # => ConversionError: Missing values for 'age' and 'name'
u = User.create(age=99, name="John Snow")
u = User.create(age=99, name=DEFAULT)
print u.name  # => None
```

This is completely backward-compatible because the new semantic is only in effect with the new class method.

**Comment by bintoro** [comment]
@edc Yeah, something like this could be useful... But it would entail giving up the nice link between `partial`, `apply_defaults`, and `init_values`, which is at the root of the issue.

If the `required` check must stay in `import_loop`, the current behavior could be emulated by `partial=True, apply_defaults=True, init_values=True`.

But then changing this setup to something actually workable for partial models would require reversing both `apply_defaults` and `init_values`. This is where the classmethod would come into play; it would have a fixed configuration for the use case.

I dunno. If there are strong reasons why the required check should happen during import, then this plan might be the way forward.

Conceptually I think the validation stage would be a better fit for enforcing field requirements. Judging by the issues raised here, many are confused about the difference between conversion and validation errors anyway, and applications are going to perform both.

How many things will break if the import step stops erroring on missing fields — I have no idea. But I have a feeling the separation between the conversion and validation checks is considered more of an annoyance than helpful, and in practice people just run both and collect the errors.

Anyway, I already tried moving the required check behind `validate()`. It works great and eliminated a whole bunch of failing tests. There's only one test that specifically checks if `required` is enforced during import, and it looks like I'm the one who wrote it...

**Comment by edc** [comment]
I might not have a full picture of the issues at play, but based on my limited understanding, I feel the solution presented here is a bit heavy. My earlier comment was more of a selfish attempt to isolate backward-incompatible changes so my work does not get interrupted. I've since given it a little more thought.

Please correct me if I am wrong. I think there are two separate issues:

1. The concept of "partial". Using a Model to represent a partial is the culprit of many issues such as unexpected/undesired behavior about default values. Things get weird when Model tries to serve two distinct purposes.

2. The semantic of setting `None` value on a field. Currently, it means to remove any existing value on a field, and the immediately side effect of falling back to default on the field.

I tend to think these are two separate issues that require attention to their root causes. Putting more options to the constructor or `import_loop` may stop the bleeding, at the cost of unnecessary complexity, IMHO.

What do think about these ideas?

1. Make Partial a first class citizen separate from the Model. When a Model class is created, automatically create a sibling class for the Partial. For example, if a `Customer` model class is created, `Customer.Partial` is the class to be used to work with partial data. Partial obviously does not do validation or anything beyond converting data, and does not respect default values, etc. It shares schema definition with its mother class, and nothing more than that. I believe its use case is limited to parsing HTTP PATCH data.

2. Cleaning up the behavior of field default values, by modeling it after structs in static typing languages like C and [Swift](https://developer.apple.com/library/prerelease/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Initialization.html). (I think a schema is basically a struct in Python.) In these languages, a default is *only* used during initialization/instantiation. After an object has been constructed and initialized, default values on fields are never used. I believe this fits our mental model much better and I would be surprised if any tests will break under this model.

**Comment by bintoro** [comment]
Yes, you're right about the funny behavior of default values. The import process always converts a `None` into the default if there is one. That doesn't make sense and will be sorted out. 

As for processing partial instances, the goal is to be able to make a distinction between
  * an explicit `None`
  * a value that has not been set.

Currently these end up being the same thing, which makes it impossible to do PATCH etc.

And it's not just about parsing incoming data. Serializing instances while leaving out unset fields must also be possible. Consequently, validating the data needs to be possible as well.

The current strategy (which certainly is up for discussion) is to simply enable missing values on instances. It's not about enabling some magical "alternative mode" but to allow an absence of values. (Traditionally, there's been a guarantee that all fields have *some* value — at least a `None`. All I'm doing is getting rid of this guarantee.)

If your application does not need to distinguish between `None` and "not set", things won't change much if at all. The compatibility issue only has to do with the import/instantiation options. See, in the current version instantiating a model

  * populates all fields with `None`
  * sets `partial=True`.

I wanted to repurpose the `partial` parameter to simultaneously act as a shortcut for a setup that makes sense for *partial instances*, but it's impossible to fully reconcile the above combination with that.

That's all this is, really.

I've gotten the branch to a point where all tests pass with minimal changes to the tests themselves. Here's how it looks:

```python
class M(Model):
    a = IntType()
    b = IntType()
```
```python
>>> m = M({'a': 1})
>>> m.a
1
>>> m.b # None
>>> m._data
{'a': 1, 'b': None}
```

See? By default it works *exactly* like now, except it wouldn't enforce `required` on the import step because I've used `partial=False` to achieve the automatic population with `None`s.

In contrast, setting `partial=True` implies `init_values=False`, and the missing field `b` won't get autofilled:

```python
>>> m = M({'a': 1}, partial=True)
>>> m.a
1
>>> m.b
Traceback (most recent call last):
  ...
schematics.exceptions.MissingValueError
>>> m._data
{'a': 1, 'b': Undefined}
>>> 
```

As shown above, trying to access a field that's not set will raise an exception as per normal Python behavior. The `MissingValueError` is a subclass of `AttributeError`, so you don't have to look for it specifically.

Internally, missing values are represented by `Undefined` placeholder objects, which preserves parity between `_data` and `_fields` and makes it easy to work with the internals since you can just iterate over everything.

In summary, the whole partial object stuff in my view is not so much about representing two different things than simply extending the capabilities of model instances. There are no two kinds of instances. You can instantiate a "regular" object and turn it into a "partial" object by saying `del obj.field`.

**Comment by edc** [comment]
Thanks for the detailed explanation, @bintoro! Really appreciated. I think we are aligned in terms of objectives and differ only in means to achieve them. I for one do not use partial, so as long as the non-partial use cases are unaffected, I am OK with the change. I still prefer to have the logic and complexity of dealing with partial taken out and reduce the number of options accepted in Model constructor, but it's probably just me.

One piece of advice, could we not expose `apply_defaults` and `init_values` as options/kwargs? I do not want to see this library getting more and more complex or requiring more explanations, and us to look at more constructor kwargs as a panacea. The possible combinations of kwargs are not fun to think about and future contributors have to perform archaeology work on these options as we did. `apply_defaults` and `init_values`  can still be made available as methods.

**Comment by bintoro** [comment]
> I still prefer to have the logic and complexity of dealing with partial taken out and reduce the number of options accepted in Model constructor

I agree about minimizing bloat in the main API.

As for the actual functionality itself, consider this example:

```python
class User(Model):
    id = IntType()
    name = StringType()
    city = StringType()
    email = EmailType()
```

Let's say there's an existing row in a database:

id   | name      | city     | email
---- | --------- | -------- | ---------
99   | bintoro   | Helsinki | kalle@...

The client sends an update:

```json
PATCH /users/99 HTTP/1.1
Content-Type: application/json

{
  "city": null
}
```

Currently, updating the record using a Schematics instance created out of the payload would reset not only the city but **every** field. To reset just the city, an application would have to manually inspect the initial data.

Alternatively, loading the old data in an instance first and using `import_data` would reset **nothing**, since the incoming `None` is not considered any different from "not present".

I believe this is a very common use case, and therefore models should be able to deal with missing values natively. Introducing a parallel construct like `PartialModel` would add enormous complexity, as you'd then have to define the interactions between them and "regular" models.

> [...] as long as the non-partial use cases are unaffected, I am OK with the change.

Yes, this is the intent. Whether "no value" should be represented as `None` or `Undefined` is completely up to the application, and can be decided on a case-by-case basis.

Some defaults are probably going to change. For example, have you noticed that running `Model.validate()` will turn all `None`s into the respective fields' defaults? I find this completely bizarre and intend to reverse the behavior (i.e., `validate()` would set `apply_defaults=False`).

> One piece of advice, could we not expose `apply_defaults` and `init_values` as options/kwargs?

That's exactly where I'm going with this. If the existing `partial` parameter could work as a shortcut for the two main cases, then the two overrides (`apply_defaults` and `init_values`) wouldn't have to appear anywhere besides the main `import_loop` function.

The way I see it, the main loops in `transforms.py` should document all available switches, while the methods leading up to them could expose only the most relevant options. We'd have a `**kwargs` slot for the more esoteric arguments.

***

I'm still not sure what to think about having the `partial` parameter control the other two. It's not without its problems. In particular, there would be a disconnect as to the meaning of `partial` between `Model.__init__()` and `Model.validate()`.

A more compatible alternative would be to leave `partial` alone and add the `init_values` parameter to the `Model` constructor. Then, `init_values` could by default dictate the `apply_defaults` setting (unless overridden), since those two definitely go hand in hand.

That way,
  * the `required` check could continue to be performed upon import
  * `partial=True` could remain the default
  * the semantics of `partial` would be preserved exactly as is
  * creating a blank instance would only require switching `init_values` to False.

I set up a parallel branch that implements this and didn't run into any surprises. I think this is a better approach. Adding one more option to `Model.__init__` is not too bad. Those who don't need it can ignore it.

**Comment by bintoro** [comment]
> This is a great example showcasing that a partial model is not a model, and using the same construct to represent them is a hack that is bound to introduce unnecessary complexity and surprising behaviors.

Fair enough. I believe the opposite — that it showcases the flaws of the *current* approach.

I mean, generic Python objects don't auto-initialize their attributes either. And neither do they equate `None` and "not set":

```python
class Foo(object):
    def __init__(self, x, y):
        self.x, self.y = x, y
```
```python
>>> obj = Foo(1, 2)
>>> obj.x
1
>>> del obj.x
>>> obj.x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Foo' object has no attribute 'x'
```

Is that also "a hack that is bound to introduce unnecessary complexity"?

I'm happy to discuss alternative ways of accomplishing the goals of the proposal, but I'm puzzled as to why gaining the ability to expressly unset a field is so terrible?

In the earlier plans, it looked like this was going to be a truly backward-incompatible change. However, now that I've warmed to the idea of keeping instances pre-populated by default, I honestly don't see what the problem is. A concrete example illustrating the issues would be great.

**Comment by edc** [comment]
Great discussion, and I appreciate all the examples.

> I'm puzzled as to why gaining the ability to expressly unset a field would be so terrible.

I don't think it's a bad idea. It would solve the problem with partial models. We would see fewer surprises. It would be a net plus to the library, and everyone would be happier.

My argument is that it *only* solves the problem with partial models, and there *might* be other ways to address it.

> generic Python objects don't auto-initialize their attributes either. And neither do they equate None and "not set":

I think the main purpose of introducing a concept of Schema into a language like Python is to have something closer to classes/structs in a traditional static language, with predefined and rigid structures and fixed "slots" for attributes. In those languages - and in other frameworks with schema like Django - you don't distinguish between `None`/`NULL` with unset slot. So the behavior of Schematics is not inherently broken. What's broken is the behavior of default value (that's for another discussion), and that we (ab)use Schematics model in supporting the concept of partial.

If we draw an analogy from between this and relational databases, which also have the concept of schema (and default and required-ness, etc), one could bend it so he can insert a partial row. It will work, but at a cost.

If we stop thinking about partial model for a minute, does the concept of unsetting a value have other values? On the other hand, is there value in making Schematics behave more like other systems with the concept of schemas (and default and required-ness), particularly from a learner perspective?

**Comment by bintoro** [comment]
> On the other hand, is there value in making Schematics behave more like other systems with the concept of schemas (and default and required-ness), particularly from a learner perspective?

Well, we already have experience with the current system, and in purely practical terms, the `None`/unset issue has been raised so many times that it's got to be addressed one way or another.

In general, the existence of a schema does not mean you *must* have a value, even a null, for every field. Relational databases do enforce this, but Schematics is a data processing library, not a persistence mechanism. It's quite normal for even schema-conforming data transmission formats to omit nonexistent values.

Is the controversy here about the entire concept of maintaining records of omitted values? Or is it mostly about how the absence of a value is exposed by the API (simulating a missing attribute)? Because that is not set in stone.

In [this comment](https://github.com/schematics/schematics/issues/263#issuecomment-150047136) I requested feedback on two alternatives I was considering. The other approach was also based on keeping tabs on `Undefined` objects in the internal `_data` structure. However, these would normally be exposed as `None`s. The "undefined" status of a field would only come into play if specifically requested:

```python
class M(Model):
    a, b, c = IntType(), IntType(), IntType()
```

```python
>>> m = M({'a': 1, 'b': None})
>>> m.a
1
>>> m.b is None
True
>>> isset(m.b)
True
>>> m.c is None
True
>>> isset(m.c)
False
>>> m.to_primitive()
{'a': 1, 'b': None, 'c': None}
>>> m.to_primitive(serialize_when_undefined=False)
{'a': 1, 'b': None}
```

**Comment by jmsdnns** [comment]
>
> Schematics is a data processing library, not a persistence mechanism
>
EXACTLY

**Comment by edc** [comment]
> the existence of a schema does not mean you must have a value, even a null, for every field. 

But isn't the semantic of null/`None`, in the context of a `IntType` field, just "there is no value here"? The only time we have to distinguish "there is no value here" and "there is no value here and ignore this when using me as a patch" is when dealing with a "partial" model.

If we don't want to rethink how partial works, then your proposal here is the way to go. But if changing how partial is done in on the table for the 2.0 release, then I believe we can get something nicer. For illustration purpose:

```python
class M(Model):
    a, b, c = IntType(), IntType(default=2), IntType()

print type(M)  # => schematics.models.ModelMeta
print type(M.Patch)  # => schematics.models.ModelPatchMeta
m = M({'a': 1})
print m.a, m.b, m.c  # => 1, 2, None
print m._data  # => {'a': 1, 'b': 2, 'c' None}. Every slot is filled with a regular model instance.
m.b = None
print m.b  # => None, because default is only applied on instantiation.

patch = M.Patch({'a': ''})
 # => ConversionError. Patch facilitates data importing and enforcing schema.
patch = M.Patch({'a': 3})
# => equivalent to the backward-compatible form: M({'a': 3}, partial=True)
print p._data  # => {'a': 1}. Patch only contains explicitly provided data.
print p.a  # => 1
print p.b  # => raise AttributeError. default value (2) is never applied to a patch
print p.c  # => raise AttributeError
p.validate()  # => TypeError: Patch does not support validation.
m.patch(patch)  # => Apply patch, recursively if patch contains nested patches
print m.a, m.b, m.c  # => 3, 2, None
```

By separating Patch into its own concept, both Patch and regular Model become easier to reason about. Patch cannot run validate, does not work with PolyModelType,  can only contain patches in its (Poly)ModelType fields, etc etc; Model always have all its attributes filled (possibly with `None`), can be patched, cannot contain patches in its (Poly)ModelType fields, etc. By assuming these differences, both are now more predictable and they won't intertwine.

**Comment by bintoro** [comment]
> But isn't the semantic of null/`None`, in the context of a `IntType` field, just "there is no value here"? The only time we have to distinguish "there is no value here" and "there is no value here and ignore this when using me as a patch" is when dealing with a "partial" model.

Correct.

I totally get your point that it would be kind of nice to be able to isolate the "unset" situation so that you don't accidentally have a mix of `None` and `Undefined` fields when you only intended to use explicit `None`s.

But the partial models *must* support validation and exporting *as is*. That is an absolute requirement; otherwise the whole thing will be half-baked. In other words, we have to be able to deal with the input

```python
{'a': 1, 'b': None}
```

and validate it and export it without merging it with a full model and having it turn into

```python
{'a': 1, 'b': None, 'c': None}
```

The idea of creating a full-blown parallel class to represent a model's partial data is likely a non-starter because all its members would have to be kept in sync with the master copies on the main class. Absolutely everything would have to implement `__getattribute__()` and intercept all operations and make sure they are performed in the parallel universe as well. Instance and subclass checks and equality comparisons would all break unless re-implemented in the metaclass.

However, I suppose in theory we *could* have a `PartialModel` class that simply delegates everything to the main model. But how, then, is this any better than simply using a regular instance of the main model and tagging it as `m.patch=True`? What does the separate class bring to the table?

***

The general idea does warrant examination, though.

That is, should we implement a `Model.patch` property that would indicate the kind of each instance? For *patch instances* the unset/`None` distinction would exist, but for *full instances* it would not. This is certainly worth considering.

```python
>>> m = M({'a': 1, 'b': None})
>>> del m.a
>>> m.a # None
>>> m.b # None
>>> m.c # None
```

```python
>>> m = M({'a': 1, 'b': None}, patch=True)
>>> del m.a
>>> m.a
MissingValueError
>>> m.b # None
>>> m.c
MissingValueError
```

Essentially, `patch` would subsume `init_values` in the constructor, but instead of only affecting instance creation, it would have a persistent effect.

The property would not be static. You could freely switch from one kind to the other. Going from `patch=True` to `patch=False` would imply the conversion of every `Undefined` to `None`.

To me, this seems pretty reasonable with no obvious drawbacks. It might even end up delivering some benefits beyond simply guarding against accidental mixing of different concepts of "no value".

What do you think?

**Comment by bintoro** [comment]
> no obvious drawbacks

> might even end up delivering some benefits

I've now played around with the idea, and both of these predictions have failed to come true. I'm not seeing any point to the exercise.

Obviously, a tree of models should consist of either full or patch instances exclusively. Making sure this is always the case is actually kind of tedious. It makes all recursive operations more complicated and precarious.

I'm trying to make the codebase easier to understand, not more complex. Unless there's a tangible, demonstrable benefit to separating full instances from patch instances, I'm not going to do it.

In the current plan, unset fields won't emerge on their own. The developer must either choose to forgo populating a new instance or explicitly `del` a value. If you do neither, there won't be any undefineds. And even if you do end up with a mixed/unknown situation, you can always

* run the import loop and have it populate the missing values; or
* export with `export_level=ALL`, which will treat missing values as if they were `None`s; or
* use `instance.get(fieldname)` to retrieve individual values, which will guarantee at least a `None`.

That's plenty of opportunity to "enforce the schema".

**Comment by jmsdnns** [comment]
Glad you drew this conclusion. I was watching some of the discussion
thinking I just don't understand the use case yet.

Originally, the whole idea with partial validation was that only the listed
fields would be validated and it mostly wouldn't matter what the
requirements were. By nature of being partial, it wasn't going to be
treated as tho the whole model would or should be there.

That's all. The output was just the validated form of whatever went in.

going much further doesn't seem like a concern schematics should have. It
has already validated whatever it could.

Further, reflecting on the term "patch" makes me think the nature of
schematics is being twisted. I am already not sure what to think of holding
on to the initial values, but some have argued for it. Now we're talking
about holding on to versions.

If a pipe line of data transformation is desireable, then perhaps we could
start a new project on the schematics github. I think it is getting further
and further away from Schematics's core competency tho and that worries me.

On Thu, Dec 3, 2015 at 10:46 AM, Kalle Tuure <notifications@github.com>
wrote:

> no obvious drawbacks
>
> might even end up delivering some benefits
>
> I've now played around with the idea, and both of these predictions have
> failed to come true. I'm not seeing any point to the exercise.
>
> Obviously, a tree of models should consist of either full or patch
> instances exclusively. Making sure this is always the case is actually kind
> of tedious. It makes all recursive operations more complicated and
> precarious.
>
> I'm trying to make the codebase easier to understand, not more complex.
> Unless there's a tangible, demonstrable benefit to separating full
> instances from patch instances, I'm not going to do it.
>
> In the current plan, unset fields won't emerge on their own. The developer
> must either choose to forgo populating a new instance or explicitly del a
> value. If you do neither, there won't be any undefineds. And even if you do
> end up with a mixed/unknown situation, you can always
>
>    - run the import loop and have it populate the missing values; or
>    - export with export_level=ALL, which will treat missing values as if
>    they were Nones; or
>    - use instance.get(fieldname) to retrieve individual values, which
>    will guarantee at least a None.
>
> That's plenty of opportunity to "enforce the schema".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/schematics/schematics/issues/263#issuecomment-161682330>
> .
>



-- 
jms dnns
me <http://jmsdnns.com> // facebook <https://www.facebook.com/jmsdnns> //
github <https://github.com/jmsdnns/> // twitter
<https://twitter.com/jmsdnns>

**Comment by bintoro** [comment]
The word "patch" could still be useful as a convenience parameter in `Model.__init__()` that says: "leave missing fields as `Undefined` and don't apply defaults either". Handling an HTTP PATCH request is pretty much the quintessential use case for doing just that.

At first I wanted to use `partial` for this, but I've backtracked because of compatibility issues.

**Comment by jmsdnns** [comment]
> The word "patch" could still be useful as a convenience parameter in
> Model.__init__() that says: "leave missing fields as Undefined and don't
> apply defaults either". Handling an HTTP PATCH request is pretty much the
> quintessential use case for doing just that.
>
If this is not the current behavior, then I agree and think this is how it
should be.

That's how it was once upon a time, but you know... I stepped away from
Schematics for some years.

**Comment by edc** [comment]
> By nature of being partial, it wasn't going to be treated as tho the whole model would or should be there. ... going much further doesn't seem like a concern schematics should have

+1. I am in favor of simplification.

In my experience, the whole partial thing is unnecessary. I have good success with something as simple as

```python
patched = dict_merge(unpatched_model.to_native(), patch_dict)
patched_model = Model(patched)
patched_model.validate()
```

> That's how it was once upon a time, but you know... I stepped away from
Schematics for some years.

Looks like we need a BDFL :)

**Comment by bintoro** [comment]
@edc Your example assumes that you want to end up with a fully populated instance. Again, wanting to export a model *while omitting some fields completely* is a very legitimate use case. Also, it's more efficient when you can just do `obj.import_data(more_data)` instead of exporting, merging the dicts, and importing again.

**Comment by j2labs** [comment]
>
> Looks like we need a BDFL :)
>
hah.

On the other hand, I've now been a VC and built some iOS apps.

Krace was BDFN (n is "now") but Kalle's been shipping the most lately. I'm
happy to watch and speak up, but history is written by the committers. :)

**Comment by edc** [comment]
> it's more efficient when you can just do obj.import_data(more_data) instead of exporting, merging the dicts, and importing again.

True. My use case requires recursive patching, and hence the export-merge-import dance. For shallow patching, a la `import_data`, one can just do `model.import_data(dict(model._data, **patch_dict))`.

> Again, wanting to export a model while omitting some fields completely is a very legitimate use case.

Don't we have `roles` for that?

> Your example assumes that you want to end up with a fully populated instance

True. My assumption is that patch by itself is not very useful.

**Comment by bintoro** [comment]
@edc Fixing `import_data` is in the pipeline. Obviously it *should* support recursive merging. By far the most efficient way to do it is during the import loop, not in two separate passes. The `import_data` method will likely just become a wrapper that delegates to `import_loop` with parameters that say "import this on top of that".

> Again, wanting to export a model while omitting some fields completely is a very legitimate use case.
>
>> Don't we have roles for that?

Roles can be used to omit fields based on their names. What if you want to omit fields that don't have a value but preserve those that contain an explicit `None`? That's not possible today.

**Comment by edc** [comment]
> What if you want to omit fields that don't have a value but preserve those that contain an explicit None?

I feel we are going in circles :)

To recap, I think we agree on the point that it's only with the patches that we need to distinguish between None and unset, and not having the ability to distinguish between the two in a patch is bad. Our disagreement is how to make patches behave more correctly. Your proposal is to fix it under the current existing framework. My suggestion is to be more aggressive and remove the support of patch from Schematics Model and move it either to a corner or to a separate library. I think my example shows that a plain old dict is a good enough representation of a patch. I'd be interested to know any use case where this approach would fail.

If you think losing support of `patch` is a big no-no, then I think your approach is quite reasonable.

**Comment by bintoro** [comment]
> I feel we are going in circles :)

Yes! You keep arguing against supporting unset fields but haven't provided a single example of the supposed problems it would introduce. Unless you can clearly articulate the issues you're concerned about, you're not going to change my view.

> I'd be interested to know any use case where this approach would fail.

Once again, it fails in validating and re-exporting the data.

If validating the data implies losing the unset/`None` distinction, how are you going to export only the fields that were originally present? Of course you can resort to storing that information elsewhere and doing the export in two passes, but that's beside the point.

Furthermore, it's inconvenient not to be able to simply store all incoming datasets as models. Who cares whether they're complete or "patches"? It's annoying to have to store partial datasets as ordinary dictionaries until you've merged them with a full model. If you're dealing with partial data, chances are you never even *want* to merge them with a full model. I may just need to validate the data and use it to update a record in a persistence layer with absolutely no regard to what values may or may not have existed previously.
